### ISSUE TITLE
Random exception handling

### ISSUE THREAD

**Issue by mihai-caraman-hs** [grpc/grpc#6429]
### Need:
Catch service errors in the client to treat them

### Deliverable:
Service exceptions consistently show in the client

### How to reproduce:
1. Have grpcio==0.13.0 and protobuf==3.0.0b2.post2
1. Copy the below proto definition in proof.proto:

    ```proto
syntax = "proto3";
package proof;
service Proof{
   rpc get (ProofRequest) returns (ProofResponse) {}
}
message ProofRequest {}
message ProofResponse {string response = 1;}
    ```

2. Run the proto generation:

    ```
protoc --python_out=. --grpc_out=. --plugin=protoc-gen-grpc=`which grpc_python_plugin` proof.proto
    ```

3. Copy the code below in proof.py:

    ```python
import time

from grpc.beta import implementations
from grpc.beta.interfaces import StatusCode

import proof_pb2

PORT = 55003

class Servicer(proof_pb2.BetaProofServicer):
     def get(self, request, context):
        try:
            raise Exception('OOPS')
            return proof_pb2.ProofResponse(response="OK")
        except Exception as e:
            context.details(str(e))
            context.code(StatusCode.INTERNAL)
            return proof_pb2.ProofResponse(response="NOT OK")

def client():
     channel = implementations.insecure_channel('localhost', PORT)
     stub = proof_pb2.beta_create_Proof_stub(channel)
     print stub.get(proof_pb2.ProofRequest(), 30)


if __name__ == '__main__':
     server = proof_pb2.beta_create_Proof_server(Servicer())
     server.add_insecure_port('[::]:%s' % PORT)
     server.start()

     try:
        while True:
            time.sleep(10)
     except KeyboardInterrupt:
        server.stop(0)
```

4. Run the server:

    ```python proof.py```

5. Run the client:
    ```
python
import proof
proof.client()
proof.client()
proof.client()
proof.client()
```
    ```
# sometimes you'll see a NOT OK response
# other times you'll see a stacktrace with OOPS (expected behavior)
>>> proof.client()
response: "NOT OK"

>>> proof.client()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "proof.py", line 23, in client
    print stub.get(proof_pb2.ProofRequest(), 30)
  File "/usr/local/lib/python2.7/dist-packages/grpc/framework/crust/implementations.py", line 73, in __call__
    protocol_options, metadata, request)
  File "/usr/local/lib/python2.7/dist-packages/grpc/framework/crust/_calls.py", line 109, in blocking_unary_unary
    return next(rendezvous)
  File "/usr/local/lib/python2.7/dist-packages/grpc/framework/crust/_control.py", line 412, in next
    raise self._termination.abortion_error
grpc.framework.interfaces.face.face.NetworkError: NetworkError(code=StatusCode.INTERNAL, details="OOPS")
    ```

**Comment by mihai-caraman-hs** [comment]
Did some digging and found out that:
 - When the OOPS exception is caught on the client, it also hangs
 - If we remove the entire line `return proof_pb2.ProofResponse(response="NOT OK")` and let it return None, it will always show the OOPS exception (and as previously stated, it will always hang)

 - When the NOT OK response was received:
```python
# list of tickets from framework/core/_operation.py:_EasyOperation.handle_ticket:
Ticket(operation_id=UUID('dc3839d3-ad7d-452a-b3e7-a3c9c95edbfc'), sequence_number=0, group=None, method=None, subscription=None, timeout=None, allowance=None, initial_metadata=None, payload=None, terminal_metadata=None, code=None, message=None, termination=None, protocol=Protocol(kind=<Kind.INVOCATION_CONTEXT: 'invocation context'>, value=<grpc._links.invocation._Context object at 0x1c6e390>))
Ticket(operation_id=UUID('dc3839d3-ad7d-452a-b3e7-a3c9c95edbfc'), sequence_number=1, group=None, method=None, subscription=<Subscription.FULL: 'full'>, timeout=None, allowance=None, initial_metadata=<grpc._cython.cygrpc.Metadata object at 0x1c2caa0>, payload=None, terminal_metadata=None, code=None, message=None, termination=None, protocol=None)
Ticket(operation_id=UUID('dc3839d3-ad7d-452a-b3e7-a3c9c95edbfc'), sequence_number=2, group=None, method=None, subscription=None, timeout=None, allowance=None, initial_metadata=None, payload=response: "NOT OK"
, terminal_metadata=None, code=None, message=None, termination=None, protocol=None)
response: "NOT OK"

#stacktrace before calling callbacks
> /usr/local/lib/python2.7/dist-packages/grpc/framework/core/_termination.py(134)_terminate_internal_only()
-> self._pool.submit(
> /usr/local/lib/python2.7/dist-packages/grpc/framework/core/_termination.py(140)_terminate_and_notify()
-> self._terminate_internal_only(outcome)
> /usr/local/lib/python2.7/dist-packages/grpc/framework/core/_termination.py(201)abort()
-> self._terminate_and_notify(outcome)
> /usr/local/lib/python2.7/dist-packages/grpc/framework/core/_operation.py(82)abort()
-> self._termination_manager.abort(outcome)
> /usr/local/lib/python2.7/dist-packages/grpc/framework/core/_end.py(72)_abort()
-> operation.abort(base.Outcome.Kind.LOCAL_SHUTDOWN)
> /usr/local/lib/python2.7/dist-packages/grpc/framework/core/_end.py(162)stop()
-> _abort(self._cycle.operations.values())
> /usr/local/lib/python2.7/dist-packages/grpc/beta/_stub.py(109)down()
-> self._end_link.stop(0).wait()
> /usr/local/lib/python2.7/dist-packages/grpc/beta/_stub.py(80)__del__()
-> self._down()
> /home/django/web/frontend/x(2)<module>()
-> import server;server.client()
*** Oldest frame
```
 - When the OOPS response was received:
```python
# list of tickets from framework/core/_operation.py:_EasyOperation.handle_ticket
Ticket(operation_id=UUID('31a55c28-0a59-4b47-a426-514febff6dd4'), sequence_number=0, group=None, method=None, subscription=None, timeout=None, allowance=None, initial_metadata=None, payload=None, terminal_metadata=None, code=None, message=None, termination=None, protocol=Protocol(kind=<Kind.INVOCATION_CONTEXT: 'invocation context'>, value=<grpc._links.invocation._Context object at 0x277f390>))
Ticket(operation_id=UUID('31a55c28-0a59-4b47-a426-514febff6dd4'), sequence_number=1, group=None, method=None, subscription=<Subscription.FULL: 'full'>, timeout=None, allowance=None, initial_metadata=<grpc._cython.cygrpc.Metadata object at 0x273daa0>, payload=None, terminal_metadata=None, code=None, message=None, termination=None, protocol=None)
Ticket(operation_id=UUID('31a55c28-0a59-4b47-a426-514febff6dd4'), sequence_number=2, group=None, method=None, subscription=None, timeout=None, allowance=None, initial_metadata=None, payload=None, terminal_metadata=<grpc._cython.cygrpc.Metadata object at 0x273daf0>, code=<StatusCode.INTERNAL: 13>, message='OOPS', termination=<Termination.TRANSMISSION_FAILURE: 'transmission failure'>, protocol=None)

#stacktrace before calling callbacks:
> /usr/local/lib/python2.7/dist-packages/grpc/framework/core/_termination.py(134)_terminate_internal_only()
-> self._pool.submit(
> /usr/local/lib/python2.7/dist-packages/grpc/framework/core/_termination.py(140)_terminate_and_notify()
-> self._terminate_internal_only(outcome)
> /usr/local/lib/python2.7/dist-packages/grpc/framework/core/_termination.py(201)abort()
-> self._terminate_and_notify(outcome)
> /usr/local/lib/python2.7/dist-packages/grpc/framework/core/_reception.py(88)_abort()
-> self._termination_manager.abort(outcome)
> /usr/local/lib/python2.7/dist-packages/grpc/framework/core/_reception.py(155)receive_ticket()
-> _utilities.Outcome(outcome_kind, ticket.code, ticket.message))
> /usr/local/lib/python2.7/dist-packages/grpc/framework/core/_operation.py(76)handle_ticket()
-> self._reception_manager.receive_ticket(ticket)
> /usr/local/lib/python2.7/dist-packages/grpc/framework/core/_end.py(204)accept_ticket()
-> operation.handle_ticket(ticket)
> /usr/local/lib/python2.7/dist-packages/grpc/framework/foundation/relay.py(92)_spin()
-> behavior(value)
> /usr/local/lib/python2.7/dist-packages/grpc/framework/foundation/logging_pool.py(41)_wrapping()
-> return behavior(*args, **kwargs)
> /usr/local/lib/python2.7/dist-packages/concurrent/futures/thread.py(61)run()
-> result = self.fn(*self.args, **self.kwargs)
> /usr/local/lib/python2.7/dist-packages/concurrent/futures/thread.py(73)_worker()
-> work_item.run()
> /usr/lib/python2.7/threading.py(504)run()
-> self.__target(*self.__args, **self.__kwargs)
> /usr/lib/python2.7/threading.py(551)__bootstrap_inner()
-> self.run()
> /usr/lib/python2.7/threading.py(524)__bootstrap()
-> self.__bootstrap_inner()
*** Oldest frame
```

Since that hanging always happens when OOPS is shown, we'll return None and keep an eye on memory leaks.

But I have to say, this is an important problem to solve before grpc would be production ready.

**Comment by nathanielmanistaatgoogle** [comment]
This is now fixed on source control master (or if it isn't, please reopen this issue) and the fix will be released on 0.15.
